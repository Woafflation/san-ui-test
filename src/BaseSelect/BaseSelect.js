import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

export const toggleSingle = (state, newSelection) => [newSelection]

export const toggleMultiple = (state, newSelection) =>
  state.selectedIndexes.includes(newSelection)
    ? state.selectedIndexes.filter(index => index !== newSelection)
    : [...state.selectedIndexes, newSelection]

class BaseSelect extends Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.string).isRequired,
    classNames: PropTypes.shape({
      base: PropTypes.string.isRequired,
      selected: PropTypes.string.isRequired
      /* disabled: PropTypes.string.isRequired */
    }).isRequired,
    /* selectedClassName: PropTypes.string, */
    selectedIndexes: PropTypes.arrayOf(PropTypes.any.isRequired),
    disabledIndexes: PropTypes.arrayOf(PropTypes.any.isRequired),
    onSelect: PropTypes.func,
    stateReducer: PropTypes.func.isRequired
  }

  static defaultProps = {
    selectedIndexes: [],
    disabledIndexes: [],
    onSelect: () => {}
    /* selectedClassName: '' */
  }

  state = {
    selectedIndexes: this.props.selectedIndexes
  }

  onClick = selectedIndex => {
    const { stateReducer, onSelect } = this.props
    this.setState(
      state => ({ selectedIndexes: stateReducer(state, selectedIndex) }),
      () => onSelect(selectedIndex, this.state)
    )
  }

  render() {
    const { classNames, options, disabledIndexes } = this.props
    const { selectedIndexes } = this.state

    return options.map(option => {
      const content = option.content || option
      const index = option.index || option
      const isDisabled = disabledIndexes.includes(index)

      return (
        <button
          key={index}
          className={
            cx({
              [classNames.base]: true,
              [classNames.selected]: selectedIndexes.includes(index),
              [classNames.disabled]: isDisabled
            })
            /* `${className} ${key === selectedIndex && */
            /* selectedClassName}` */
          }
          onClick={!isDisabled ? () => this.onClick(index) : undefined}
        >
          {content}
        </button>
      )
    })
  }
}

export default BaseSelect
