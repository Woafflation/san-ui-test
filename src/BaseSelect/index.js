export { default, toggleSingle, toggleMultiple } from './BaseSelect'
