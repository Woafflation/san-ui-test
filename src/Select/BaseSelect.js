import React, { Component } from 'react'
import PropTypes from 'prop-types'

class BaseSelect extends Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.string).isRequired,
    className: PropTypes.string.isRequired,
    selectedClassName: PropTypes.string,
    selectedIndex: PropTypes.number,
    onSelect: PropTypes.func
  }

  static defaultProps = {
    selectedIndex: 0,
    onSelect: () => {},
    selectedClassName: ''
  }

  state = {
    selectedIndex: this.props.selectedIndex
  }

  onClick = selectedIndex => {
    const { onSelect } = this.props
    this.setState({ selectedIndex })

    onSelect(selectedIndex)
  }

  render() {
    const { className, selectedClassName, options } = this.props
    const { selectedIndex } = this.state
    return options.map((option, i) => {
      return (
        <button
          key={i}
          className={`${className} ${i === selectedIndex && selectedClassName}`}
          onClick={() => this.onClick(i)}
        >
          {option}
        </button>
      )
    })
  }
}

export default BaseSelect
