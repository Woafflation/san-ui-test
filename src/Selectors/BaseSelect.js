import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

export const toggleSingle = (state, newSelection) => [newSelection]

export const toggleMultiple = (state, newSelection) =>
  state.selectedIndexes.includes(newSelection)
    ? state.selectedIndexes.filter(index => index !== newSelection)
    : [...state.selectedIndexes, newSelection]

class BaseSelect extends Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.string).isRequired,
    className: PropTypes.string.isRequired,
    selectedClassName: PropTypes.string.isRequired,
    disabledClassName: PropTypes.string.isRequired,
    selectedIndexes: PropTypes.arrayOf(PropTypes.any.isRequired),
    disabledIndexes: PropTypes.arrayOf(PropTypes.any.isRequired),
    onSelect: PropTypes.func,
    stateReducer: PropTypes.func.isRequired
  }

  static defaultProps = {
    selectedIndexes: [],
    disabledIndexes: [],
    onSelect: () => {}
    /* selectedClassName: '' */
  }

  state = {
    selectedIndexes: this.props.selectedIndexes
  }

  onClick = selectedIndex => {
    const { stateReducer, onSelect } = this.props
    this.setState(
      state => ({ selectedIndexes: stateReducer(state, selectedIndex) }),
      () => onSelect(selectedIndex, this.state)
    )
  }

  render() {
    const {
      className,
      selectedClassName,
      disabledClassName,
      options,
      disabledIndexes
    } = this.props
    const { selectedIndexes } = this.state

    return options.map(option => {
      const content = option.content || option
      const index = option.index || option
      const isDisabled = disabledIndexes.includes(index)

      return (
        <div
          key={index}
          className={
            cx({
              [className]: true,
              [selectedClassName]: selectedIndexes.includes(index),
              [disabledClassName]: isDisabled
            })
            /* `${className} ${key === selectedIndex && */
            /* selectedClassName}` */
          }
          onClick={!isDisabled ? () => this.onClick(index) : undefined}
        >
          {content}
        </div>
      )
    })
  }
}

export default BaseSelect
