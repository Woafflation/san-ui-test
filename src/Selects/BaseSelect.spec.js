/* eslint-env jest */
import React from 'react'
import toJson from 'enzyme-to-json'
import { shallow, mount } from 'enzyme'
import BaseSelect, { toggleSingle, toggleMultiple } from './BaseSelect'

const defaultOptions = ['1w', '1m', '3m', '6m', '1y', 'all']
const simpleReducer = (state, newSelection) => [newSelection]
const classNames = {
  className: '',
  selectedClassName: '',
  disabledClassName: ''
}

describe('BaseSelect component', () => {
  let mockCb

  beforeEach(() => {
    mockCb = jest.fn()
  })

  it('smoke', () => {
    const wrapper = shallow(
      <BaseSelect
        stateReducer={simpleReducer}
        options={defaultOptions}
        onSelect={mockCb}
        {...classNames}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should call onSelect when option is clicked', () => {
    const wrapper = mount(
      <BaseSelect
        stateReducer={simpleReducer}
        options={defaultOptions}
        onSelect={mockCb}
        {...classNames}
      />
    )

    wrapper
      .find('div')
      .at(2)
      .simulate('click')
    expect(mockCb).toHaveBeenCalled()
  })

  it('should contain selected option specified by the default', () => {
    const wrapper = mount(
      <BaseSelect
        stateReducer={simpleReducer}
        options={defaultOptions}
        onSelect={mockCb}
        defaultSelectedIndexes={defaultOptions.slice(0, 1)}
        {...classNames}
      />
    )

    expect(wrapper.state().selectedIndexes).toEqual(defaultOptions.slice(0, 1))
  })

  describe('toggleSingle state reducer', () => {
    it('should update state with clicked option', () => {
      const wrapper = mount(
        <BaseSelect
          stateReducer={toggleSingle}
          options={defaultOptions}
          onSelect={mockCb}
          {...classNames}
        />
      )
      const CLICKED_OPTION_INDEX = 3

      wrapper
        .find('div')
        .at(CLICKED_OPTION_INDEX)
        .simulate('click')

      expect(wrapper.state().selectedIndexes).toEqual(
        defaultOptions.slice(CLICKED_OPTION_INDEX, CLICKED_OPTION_INDEX + 1)
      )
    })

    it('state should contain only 1 last selected option after multiple clicks', () => {
      const wrapper = mount(
        <BaseSelect
          stateReducer={toggleSingle}
          options={defaultOptions}
          onSelect={mockCb}
          {...classNames}
        />
      )
      const OPTIONS_INDEX_CLICK_SEQUENCE = [1, 4, 3]
      const LAST_CLICKED_INDEX =
        OPTIONS_INDEX_CLICK_SEQUENCE[OPTIONS_INDEX_CLICK_SEQUENCE.length - 1]

      OPTIONS_INDEX_CLICK_SEQUENCE.forEach(index =>
        wrapper
          .find('div')
          .at(index)
          .simulate('click')
      )

      expect(wrapper.state().selectedIndexes).toEqual(
        defaultOptions.slice(LAST_CLICKED_INDEX, LAST_CLICKED_INDEX + 1)
      )
    })
  })

  describe('toggleMultiple state reducer', () => {
    it('should update state with clicked option', () => {
      const wrapper = mount(
        <BaseSelect
          stateReducer={toggleMultiple}
          options={defaultOptions}
          onSelect={mockCb}
          {...classNames}
        />
      )
      const CLICKED_OPTION_INDEX = 3

      wrapper
        .find('div')
        .at(CLICKED_OPTION_INDEX)
        .simulate('click')

      expect(wrapper.state().selectedIndexes).toEqual(
        defaultOptions.slice(CLICKED_OPTION_INDEX, CLICKED_OPTION_INDEX + 1)
      )
    })

    it('state should contain all selected options after multiple clicks', () => {
      const wrapper = mount(
        <BaseSelect
          stateReducer={toggleMultiple}
          options={defaultOptions}
          onSelect={mockCb}
          {...classNames}
        />
      )
      const OPTIONS_INDEX_CLICK_SEQUENCE = [1, 4, 3]

      OPTIONS_INDEX_CLICK_SEQUENCE.forEach(index =>
        wrapper
          .find('div')
          .at(index)
          .simulate('click')
      )

      expect(wrapper.state().selectedIndexes).toEqual(
        OPTIONS_INDEX_CLICK_SEQUENCE.map(index => defaultOptions[index])
      )
    })
  })
})
