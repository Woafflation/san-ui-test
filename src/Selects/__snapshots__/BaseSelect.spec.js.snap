// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSelect component smoke 1`] = `
Array [
  <div
    className="wrapper "
    key="1w"
    onClick={[Function]}
  >
    1w
  </div>,
  <div
    className="wrapper "
    key="1m"
    onClick={[Function]}
  >
    1m
  </div>,
  <div
    className="wrapper "
    key="3m"
    onClick={[Function]}
  >
    3m
  </div>,
  <div
    className="wrapper "
    key="6m"
    onClick={[Function]}
  >
    6m
  </div>,
  <div
    className="wrapper "
    key="1y"
    onClick={[Function]}
  >
    1y
  </div>,
  <div
    className="wrapper "
    key="all"
    onClick={[Function]}
  >
    all
  </div>,
]
`;
